### PR TITLE
Introduce LS_JAVA_HOME environment variable

### PR DESCRIPTION
**Pull request by andsel**
## Release notes

Logstash's  launch script support `LS_JAVA_HOME` as an override of environment variable over `JAVA_HOME`. 
It's the path to an installed JDK to be used to launch Logstash's service and support tools.
`JAVA_HOME` is going to be removed, however if `LS_JAVA_HOME` is not specified then Logstash still looks for `JAVA_HOME`.

## What does this PR do?

This PR modifies the launch scripts to take care of the `LS_JAVA_HOME` giving precedence over the `JAVA_HOME`, which is still used it the first is not found.

## Why is it important/What is the impact to the user?

The existing `JAVA_HOME` is an environment variable globally used by all JVM applications present on a host.
With the introduction of `LS_JAVA_HOME` the user is able to customize the JDK used by Logstash without changing the global `JAVA_HOME`. It introduces a layer of separation, permitting a more fine grained control of the JDK customization for Logstash.

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] ~~I have commented my code, particularly in hard-to-understand areas~~
- [x]  have made corresponding changes to the documentation
- [ ] ~~I have made corresponding change to the default configuration files (and/or docker env variables)~~
- [ ] ~~I have added tests that prove my fix is effective or that my feature works~~

## Author's Checklist

- [x] verify it works setting `LS_JAVA_HOME` on Windows and Linux

## How to test this PR locally

-  checkout this branch on your machine
- on `bash` set the `LS_JAVA_HOME` `export LS_JAVA_HOME=/path/to/jdk` and run `bin/logstash -e "input{stdin{}} output {stdout{codec => rubydebug}}"
- on Windows cmd `set LS_JAVA_HOME="C:\path\to\JDK"`  and run `bin\logstash -e "input{stdin{}} output {stdout{codec => rubydebug}}"

In both cases it prints the the JDK used and must correspond to what was selected, verify also with rest API `curl -XGET 'localhost:9600/_node/jvm?pretty'`

## Related issues

-

### CODE REVIEW

File: bin/logstash.lib.sh [hunk 1]
```diff
@@ -90,7 +91,17 @@ setup_java() {
   if [ -z "$JAVACMD" ]; then
     setup_bundled_jdk_part
     JAVACMD_TEST=`command -v java`
-    if [ -n "$JAVA_HOME" ]; then
+    if [ -n "$LS_JAVA_HOME" ]; then
+      echo "Using LS_JAVA_HOME defined java: ${LS_JAVA_HOME}"
+      if [ -x "$LS_JAVA_HOME/bin/java" ]; then
+        JAVACMD="$LS_JAVA_HOME/bin/java"
+        if [ -d "${LOGSTASH_HOME}/${BUNDLED_JDK_PART}" -a -x "${LOGSTASH_HOME}/${BUNDLED_JDK_PART}/bin/java" ]; then
+          echo "WARNING, using LS_JAVA_HOME while Logstash distribution comes with a bundled JDK"
```
Comment (roaksoax) [inline bin/logstash.lib.sh:99]: I believe the format for warnings is "WARNING: <message>". You are using a ',' instead. So should be:

WARNING: Using LS_JAVA_HOME env variable while Logstash distribution comes with a bundled JDK (<version>)"

I made a couple of minor adjustments. Would be nice to specify which version of the JDK is bundled (the full version). Lets use end of sentence punctuation here as well.

File: bin/logstash.lib.sh [hunk 2]
```diff
@@ -112,7 +123,7 @@ setup_java() {
   fi
 
   if [ ! -x "$JAVACMD" ]; then
-    echo "could not find java; set JAVA_HOME or ensure java is in PATH"
+    echo "could not find java; set LS_JAVA_HOME or ensure java is in PATH"
```
Comment (roaksoax) [inline bin/logstash.lib.sh:126]: Should the first letter be upper cased? (It is a new sentence(. Also, I believe we should use proper punctuation, so should likely need a '.' at the end of all sentences.

File: bin/logstash.lib.sh [hunk 3]
```diff
@@ -90,7 +91,17 @@ setup_java() {
...
-    if [ -n "$JAVA_HOME" ]; then
+    if [ -n "$LS_JAVA_HOME" ]; then
+      echo "Using LS_JAVA_HOME defined java: ${LS_JAVA_HOME}"
+      if [ -x "$LS_JAVA_HOME/bin/java" ]; then
+        JAVACMD="$LS_JAVA_HOME/bin/java"
+        if [ -d "${LOGSTASH_HOME}/${BUNDLED_JDK_PART}" -a -x "${LOGSTASH_HOME}/${BUNDLED_JDK_PART}/bin/java" ]; then
+          echo "WARNING, using LS_JAVA_HOME while Logstash distribution comes with a bundled JDK"
+        fi
+      else
+        echo "Invalid LS_JAVA_HOME, doesn't contain bin/java executable"
```
Comment (roaksoax) [inline bin/logstash.lib.sh:102]: missing end of sentence punctuation.

File: bin/setup.bat [hunk 4]
```diff
@@ -20,7 +20,13 @@ for %%I in ("%LS_HOME%..") do set LS_HOME=%%~dpfI
 
 rem ### 2: set java
 
-if defined JAVA_HOME (
+if defined LS_JAVA_HOME (
+  set JAVA="%LS_JAVA_HOME%\bin\java.exe"
+  echo Using LS_JAVA_HOME defined java: %LS_JAVA_HOME%
+  if exist "%LS_HOME%\jdk" (
+    echo WARNING, using LS_JAVA_HOME while Logstash distribution comes with a bundled JDK
```
Comment (roaksoax) [inline bin/setup.bat:27]: Incorrect format, plus missing punctuation.

File: bin/logstash.lib.sh [hunk 5]
```diff
@@ -90,7 +91,17 @@ setup_java() {
   if [ -z "$JAVACMD" ]; then
     setup_bundled_jdk_part
     JAVACMD_TEST=`command -v java`
-    if [ -n "$JAVA_HOME" ]; then
+    if [ -n "$LS_JAVA_HOME" ]; then
+      echo "Using LS_JAVA_HOME defined java: ${LS_JAVA_HOME}"
```
Comment (roaksoax) [inline bin/logstash.lib.sh:95]: same here, punctuation?

**Review by roaksoax — APPROVED** [review]
The actual changes look good to me. I'll approve on that basis, however, could you please make sure all the comments about messages are addressed before landing?

File: bin/logstash.lib.sh (hunk 1 above)
Comment (andsel) [reply bin/logstash.lib.sh:99]: One way to print the full version of the bundled JDK could be to execute the `"${LOGSTASH_HOME}/${BUNDLED_JDK_PART}/bin/java -version" `. But this produces 3 lines of output after the "WARNING:" line, it's not inlined.

File: docs/static/jvm.asciidoc [hunk 6]
```diff
@@ -47,17 +47,19 @@ Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.1+13-LTS, mixed mode)
...
+==== `LS_JAVA_HOME` and `JAVA_HOME`
 
-{ls} uses the Java version set in `JAVA_HOME`. The `JAVA_HOME` environment
-variable must be set for {ls} to operate correctly. 
+{ls} uses the Java version set in `LS_JAVA_HOME`. The `LS_JAVA_HOME` environment
+variable must be set for {ls} to operate correctly.
 
-On some Linux systems, you may need to have the `JAVA_HOME` environment
+If {ls} doesn't find `LS_JAVA_HOME` it tries to fall back on `JAVA_HOME` but this behaviour
+could be subject to be removed.
```
Comment (karenzone) [inline docs/static/jvm.asciidoc:56]:
```suggestion
If {ls} doesn't find `LS_JAVA_HOME`, it tries to fall back to `JAVA_HOME`.  Note that this behavior is experimental and could be subject to removal in later releases.
```

File: docs/static/jvm.asciidoc [hunk 7]
```diff
@@ -47,17 +47,19 @@ Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.1+13-LTS, mixed mode)
...
+On some Linux systems, you may need to have the `LS_JAVA_HOME` environment
```
Comment (karenzone) [inline docs/static/jvm.asciidoc:57]:
```suggestion

On some Linux systems, you may need to have the `LS_JAVA_HOME` environment
```

Comment (karenzone) [reply docs/static/jvm.asciidoc:57]: Suggest return to break up text block for easier parsing.

**Review by karenzone — COMMENTED** [review]
Thanks for these doc updates, @andsel  I left some comments inline for your consideration.

**Comment by andsel** [comment]
@karenzone may I ask you a quick eye on the docs part of this PR, located in file `docs/static/jvm.asciidoc`.
This PR changes the Logstash's launch scripts to that a user can specify the customization of the JDK he wants to use also with `LS_JAVA_HOME`. This approach has the advantage to don't force the user to change the globally used `JAVA_HOME` and interfere with other JVM application he could run on the host.

**Pull request merged**